### PR TITLE
Add assemblyUnzipDirectory to AssemblyOption to use a different directory for unzipping jars

### PR DESCRIPTION
**Pull request by er1c**
Adding as a draft for just reference for now

This is a draft implementation for https://github.com/sbt/sbt-assembly/issues/445

* Use hardlinks as basis for copying from unzip cache directory when enabled
* Rebased ontop of https://github.com/sbt/sbt-assembly/pull/449

#### TODO
- [x] Add/update README.md examples
- [x] Create a flag for the hardlink feature? or have a way to explicitly turn it off (or default off)
- [x] Run through my work's CI system for extra validation

### CODE REVIEW

File: src/main/scala/sbtassembly/Assembly.scala [hunk 1]
```diff
@@ -268,24 +272,48 @@ object Assembly {
...
         {
           log.debug("Including: %s".format(jarName))
-          IO.delete(dest)
-          dest.mkdir()
-          AssemblyUtils.unzip(jar.data, dest, log)
+          // Use cache location
+          if (ao.cacheUnzip && jarNameCachePath.exists && IO.read(jarNameCachePath) == jar.data.getCanonicalPath) {
+            if (cacheDir != dest) {
+              log.debug("Copying from unzip cache: %s".format(jarName))
+              // TODO: Can we use sym or hard links?
```
Comment (er1c) [inline src/main/scala/sbtassembly/Assembly.scala:298]: @eed3si9n

File: src/sbt-test/caching/unzip/build.sbt [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+      if ( jarHash.exists )
+      {
+        val prevHash = IO.read(jarHash)
+        s.log.info( "Checking hash: " + hash + ", " + prevHash )
+        assert( hash == prevHash )
+      }
+      IO.write( jarHash, hash )
+      ()
+    },
+    TaskKey[Unit]("checkunzip") := {
```
Comment (er1c) [inline src/sbt-test/caching/unzip/build.sbt:28]: @eed3si9n I think this covers the unit tests I was thinking of, do you have any more ideas?

File: src/main/scala/sbtassembly/Assembly.scala [hunk 3]
```diff
@@ -268,24 +272,47 @@ object Assembly {
...
+              AssemblyUtils.copyDirectory(cacheDir, dest, hardLink = true)
+              IO.delete(jarNameFinalPath) // write after merge/shade rules applied
+            }
+          } else {
+            // Use cache dir
+            if (ao.cacheUnzip && jarNameFinalPath != jarNameCachePath) {
+              IO.delete(cacheDir)
+              cacheDir.mkdir()
+              AssemblyUtils.unzip(jar.data, cacheDir, log)
+              IO.copyDirectory(cacheDir, dest)
```
Comment (er1c) [inline src/main/scala/sbtassembly/Assembly.scala:307]: this should be `hardlink = true`

Comment (er1c) [inline src/main/scala/sbtassembly/Assembly.scala]: double check `hardLink = true`

**Comment by er1c** [comment]
Found one issue with the hardlinks, one of my assumptions of the caches being on the same devices is flawed :)

```
[error] java.nio.file.FileSystemException: /local/data_importer/target/streams/_global/assembly/_global/streams/assembly/2669ae58f340f420d4802cc2289f5ba407c4e2bc_a9a7907ae54d79c51d9dd7e07bc71162aa0c71a2_0289763ca8ca20fbc99cf2a3bcea50af3cd6be74/META-INF/MANIFEST.MF -> /.cache/sbt/v1/assembly-dependencies/2669ae58f340f420d4802cc2289f5ba407c4e2bc_a9a7907ae54d79c51d9dd7e07bc71162aa0c71a2_0289763ca8ca20fbc99cf2a3bcea50af3cd6be74/META-INF/MANIFEST.MF: Invalid cross-device link
[error] 	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:91)
[error] 	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
[error] 	at sun.nio.fs.UnixFileSystemProvider.createLink(UnixFileSystemProvider.java:476)
[error] 	at java.nio.file.Files.createLink(Files.java:1086)
[error] 	at sbtassembly.AssemblyUtils$.copyFile(AssemblyUtils.scala:174)
[error] 	at sbtassembly.AssemblyUtils$.copyImpl(AssemblyUtils.scala:152)
[error] 	at sbtassembly.AssemblyUtils$.$anonfun$copy$1(AssemblyUtils.scala:138)
[error] 	at scala.Function$.$anonfun$tupled$1(Function.scala:79)
```

This either needs to be caught and then just copied, or have a toggle earlier up to see if hardlink is supported

File: src/main/scala/sbtassembly/Assembly.scala [hunk 4]
```diff
@@ -219,24 +217,30 @@ object Assembly {
...
   def assembleMappings(classpath: Classpath, dependencies: Classpath,
       ao: AssemblyOption, log: Logger): Vector[MappingSet] = {
-    val tempDir = ao.assemblyDirectory.get
-    if (!ao.cacheUnzip) IO.delete(tempDir)
-    if (!tempDir.exists) tempDir.mkdir()
+    val assemblyDir = ao.assemblyDirectory.get
+    val assemblyUnzipDir = ao.assemblyUnzipDirectory.getOrElse(assemblyDir)
+
+    if (!ao.cacheOutput) IO.delete(assemblyDir)
+    if (!ao.cacheUnzip) IO.delete(assemblyUnzipDir)
```
Comment (er1c) [inline src/main/scala/sbtassembly/Assembly.scala]: This probably should only be deleted if in the local target/

File: src/main/scala/sbtassembly/Assembly.scala [hunk 5]
```diff
@@ -219,24 +217,30 @@ object Assembly {
   // which jars exactly belong to the deps for packageDependency option.
   def assembleMappings(classpath: Classpath, dependencies: Classpath,
       ao: AssemblyOption, log: Logger): Vector[MappingSet] = {
-    val tempDir = ao.assemblyDirectory.get
-    if (!ao.cacheUnzip) IO.delete(tempDir)
```
Comment (er1c) [inline src/main/scala/sbtassembly/Assembly.scala:225]: I should double check the original delete code, not sure if the `cacheOutput` was being used?

File: src/sbt-test/caching/unzip/build.sbt [hunk 6]
```diff
@@ -17,6 +24,14 @@ lazy val root = (project in file(".")).
       _.withCacheOutput(true)
         .withCacheUnzip(true)
         .withAssemblyUnzipDirectory(tempUnzipDir)
+        .withUseHardLinks(true)
+    },
+    logLevel in assembly := sbt.Level.Info,
+    extraLoggers in assembly ~= { currentFunction =>
```
Comment (er1c) [inline src/sbt-test/caching/unzip/build.sbt]: @eed3si9n do you see how this isn't working?  I just get a `List()`

Comment (eed3si9n) [reply src/sbt-test/caching/unzip/build.sbt]: We rewrote logger / went back to using our own logger so this feature may or may not work any more actually.

Comment (er1c) [reply src/sbt-test/caching/unzip/build.sbt]: Thanks for the tip, found a work around!

**Comment by er1c** [comment]
Got some corner case that I need to figure out:

```
[info] Copying shapeless_2.12-2.3.3.jar from unzip cache: /home/docker/.cache/sbt/v1/assembly/dependencies/91e75de52000c787900cf3b917ad2ffe0e5035ea_6041e2c4871650c556a9c6842e43c04ed462b11f_806e468a010b7087977fac295351e79a95d7c832, to: /shared/text_factor/scala/text_tokenizer/target/streams/_global/assembly/_global/streams/assembly/91e75de52000c787900cf3b917ad2ffe0e5035ea_6041e2c4871650c556a9c6842e43c04ed462b11f_806e468a010b7087977fac295351e79a95d7c832, for project: text-factor/text_factor
[info] Calculate mappings...
[info] Checking every *.class/*.jar file's SHA-1.
[info] Merging files...
[info] Strategy 'deduplicate' was applied to 6 files (Run the task at debug level to see details)
[info] Strategy 'discard' was applied to 417 files (Run the task at debug level to see details)
[info] Strategy 'filterDistinctLines' was applied to 6 files (Run the task at debug level to see details)
[info] Strategy 'rename' was applied to 6 files (Run the task at debug level to see details)
[info] SHA-1: d1fbb4a0902d74b3746362fdb849620f582b548e
[error] java.lang.IllegalArgumentException: requirement failed: Source file '/home/docker/.cache/sbt/v1/assembly/dependencies/346af639499c3d3ed6bb01314995d424e58af9b2_ac435f9d09ce25fc2c3dcade9b1a567466f6245a_806e468a010b7087977fac295351e79a95d7c832/META-INF/maven/com.amazonaws/aws-java-sdk-serverlessapplicationrepository/pom.properties' does not exist.
[error] 	at scala.Predef$.require(Predef.scala:281)
[error] 	at sbtassembly.AssemblyUtils$.copyFile(AssemblyUtils.scala:193)
[error] 	at sbtassembly.AssemblyUtils$.copyImpl(AssemblyUtils.scala:178)
[error] 	at sbtassembly.AssemblyUtils$.$anonfun$copy$1(AssemblyUtils.scala:164)
[error] 	at scala.Function$.$anonfun$tupled$1(Function.scala:79)
[error] 	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:286)
[error] 	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
[error] 	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
[error] 	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
[error] 	at scala.collection.TraversableLike.map(TraversableLike.scala:286)
[error] 	at scala.collection.TraversableLike.map$(TraversableLike.scala:279)
[error] 	at scala.collection.AbstractTraversable.map(Traversable.scala:108)
[error] 	at sbtassembly.AssemblyUtils$.copy(AssemblyUtils.scala:164)
[error] 	at sbtassembly.AssemblyUtils$.copyDirectory(AssemblyUtils.scala:152)
[error] 	at sbtassembly.Assembly$.$anonfun$processDependencyJars$12(Assembly.scala:462)
[error] 	at sbt.State$StateOpsImpl$$anon$1.call(State.scala:389)
[error] 	at xsbt.boot.Locks$GlobalLock.withChannel$1(Locks.scala:113)
[error] 	at xsbt.boot.Locks$GlobalLock.withChannelRetries$1(Locks.scala:91)
[error] 	at xsbt.boot.Locks$GlobalLock.$anonfun$withFileLock$1(Locks.scala:119)
```

File: src/main/scala/sbtassembly/Assembly.scala [hunk 7]
```diff
@@ -357,6 +388,183 @@ object Assembly {
...
+  ): ParVector[(File, File)] = {
+    val defaultAssemblyDir = assemblyOption.assemblyDirectory.get
+    val assemblyUnzipDir: File = assemblyOption.assemblyUnzipDirectory.getOrElse(defaultAssemblyDir)
+    val assemblyDir: Option[File] = if (isCacheOnly) None else Some(defaultAssemblyDir)
+    val isSameDir: Boolean = assemblyDir.exists{ _ == assemblyUnzipDir }
+
+    if (!assemblyUnzipDir.exists) IO.createDirectory(assemblyUnzipDir)
+    if (assemblyDir.isDefined && !assemblyDir.get.exists) IO.createDirectory(assemblyDir.get)
+
+    state.locked(assemblyUnzipDir / "sbt-assembly.lock") {
```
Comment (er1c) [inline src/main/scala/sbtassembly/Assembly.scala:421]: I really didn't want to add a lock here (I originally tried locking on individual jar file), but this seemed to be the only reliable place to put it in a very complex project.

File: src/main/scala/sbtassembly/Assembly.scala [hunk 8]
```diff
@@ -357,6 +388,183 @@ object Assembly {
...
+            r.isApplicableToAll ||
+              jarToAttributedFiles.getOrElse(jar, Vector.empty)
+                .flatMap(_.metadata.get(moduleID.key))
+                .map(m => ModuleCoordinate(m.organization, m.name, m.revision))
+                .exists(r.isApplicableTo)
+          }
+        hash = sha1name(jar) + "_" + sha1content(jar) + "_" + sha1rules(jarRules)
+        jarOutputDir = (assemblyDir.getOrElse(assemblyUnzipDir) / hash).getCanonicalFile
+      } yield {
+        // TODO: Canonical path might be problematic if mount points inside docker are different
```
Comment (er1c) [inline src/main/scala/sbtassembly/Assembly.scala:464]: This is a problem, if say you have a `/home/packer/.cache/sbt` that then gets mounted into a docker container as `/home/docker/.cache/sbt` and then get something like:

```
~/.cache/sbt/v1/assembly/dependencies $ cat fdf034b77ea4c96874cbd087f89a92d9c6169839_66a4cd59bc0daa39b67b0b5e496cee078b835d51_c3c65759ec2815d6ff9909c6e810928dba1ad3a0.jarName
/home/docker/.cache/coursier/cache/http/nexus-aws.dot-data.net/repository/maven-public/org/typelevel/spire-macros_2.12/0.17.0-M1/spire-macros_2.12-0.17.0-M1.jar
```

Trying to decide/figure out if there is an easy way to canonicalize based upon what the current cache directories are, rather fan full absolute path.

Comment (er1c) [reply src/main/scala/sbtassembly/Assembly.scala:464]: I think its safer to keep this in there, rather than do any canonicalization

File: build.sbt [hunk 9]
```diff
@@ -21,7 +21,6 @@ lazy val root = (project in file("."))
         case "2.12" => "1.2.8"
       }
     }
-    Compile / generateContrabands / sourceManaged := baseDirectory.value / "src" / "main" / "contraband-scala"
```
Comment (eed3si9n) [inline build.sbt]: Why is Contraband dropped here?

File: src/main/scala/sbtassembly/Assembly.scala [hunk 10]
```diff
@@ -43,8 +44,9 @@ object Assembly {
     import java.util.jar.{Attributes, Manifest}
 
     lazy val (ms: Vector[(File, String)], stratMapping: List[(String, MergeStrategy)]) = {
-      log.debug("Merging files...")
+      log.info("Merging files...")
```
Comment (eed3si9n) [inline src/main/scala/sbtassembly/Assembly.scala]: Could we put this back to debug?

File: build.sbt (hunk 9 above)
Comment (er1c) [reply build.sbt]: It's not used anymore, I can add it back, I had a question earlier about whether it should be used or not.  I'll ping you a link
